# What drives the price of a car?

![](images/kurt.jpeg)

**OVERVIEW**

In this application, you will explore a dataset from Kaggle. The original dataset contained information on 3 million used cars. The provided dataset contains information on 426K cars to ensure speed of processing.  Your goal is to understand what factors make a car more or less expensive.  As a result of your analysis, you should provide clear recommendations to your client -- a used car dealership -- as to what consumers value in a used car.

### CRISP-DM Framework

<center>
    <img src = images/crisp.png width = 50%/>
</center>


To frame the task, throughout our practical applications, we will refer back to a standard process in industry for data projects called CRISP-DM.  This process provides a framework for working through a data problem.  Your first step in this application will be to read through a brief overview of CRISP-DM [here](https://mo-pcco.s3.us-east-1.amazonaws.com/BH-PCMLAI/module_11/readings_starter.zip).  After reading the overview, answer the questions below.

### Business Understanding

From a business perspective, we are tasked with identifying key drivers for used car prices.  In the CRISP-DM overview, we are asked to convert this business framing to a data problem definition.  Using a few sentences, reframe the task as a data task with the appropriate technical vocabulary. 

#### Objective is to develop a linear regression model for predicting car priced based on different car features like region, year, manufacturer, condition, odometer etc. As a part of feature engineering, we need to determine the most important features impacting resultant car price.

### Data Understanding

After considering the business understanding, we want to get familiar with our data.  Write down some steps that you would take to get to know the dataset and identify any quality issues within.  Take time to get to know the dataset and explore what information it contains and how this could be used to inform your business understanding.

In [1]:
import pandas as pd

In [2]:
# Load data
df = pd.read_csv("data/vehicles.csv")

In [3]:
# Basic feature info
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 426880 entries, 0 to 426879
Data columns (total 18 columns):
 #   Column        Non-Null Count   Dtype  
---  ------        --------------   -----  
 0   id            426880 non-null  int64  
 1   region        426880 non-null  object 
 2   price         426880 non-null  int64  
 3   year          425675 non-null  float64
 4   manufacturer  409234 non-null  object 
 5   model         421603 non-null  object 
 6   condition     252776 non-null  object 
 7   cylinders     249202 non-null  object 
 8   fuel          423867 non-null  object 
 9   odometer      422480 non-null  float64
 10  title_status  418638 non-null  object 
 11  transmission  424324 non-null  object 
 12  VIN           265838 non-null  object 
 13  drive         296313 non-null  object 
 14  size          120519 non-null  object 
 15  type          334022 non-null  object 
 16  paint_color   296677 non-null  object 
 17  state         426880 non-null  object 
dtypes: f

In [4]:
# Describe data
df.describe()

id         price           year      odometer
count  4.268800e+05  4.268800e+05  425675.000000  4.224800e+05
mean   7.311487e+09  7.519903e+04    2011.235191  9.804333e+04
std    4.473170e+06  1.218228e+07       9.452120  2.138815e+05
min    7.207408e+09  0.000000e+00    1900.000000  0.000000e+00
25%    7.308143e+09  5.900000e+03    2008.000000  3.770400e+04
50%    7.312621e+09  1.395000e+04    2013.000000  8.554800e+04
75%    7.315254e+09  2.648575e+04    2017.000000  1.335425e+05
max    7.317101e+09  3.736929e+09    2022.000000  1.000000e+07

In [5]:
# Print sample data
df.head()

id                  region  price  year manufacturer model  \
0  7222695916                prescott   6000   NaN          NaN   NaN   
1  7218891961            fayetteville  11900   NaN          NaN   NaN   
2  7221797935            florida keys  21000   NaN          NaN   NaN   
3  7222270760  worcester / central MA   1500   NaN          NaN   NaN   
4  7210384030              greensboro   4900   NaN          NaN   NaN   

  condition cylinders fuel  odometer title_status transmission  VIN drive  \
0       NaN       NaN  NaN       NaN          NaN          NaN  NaN   NaN   
1       NaN       NaN  NaN       NaN          NaN          NaN  NaN   NaN   
2       NaN       NaN  NaN       NaN          NaN          NaN  NaN   NaN   
3       NaN       NaN  NaN       NaN          NaN          NaN  NaN   NaN   
4       NaN       NaN  NaN       NaN          NaN          NaN  NaN   NaN   

  size type paint_color state  
0  NaN  NaN         NaN    az  
1  NaN  NaN         NaN    ar  
2  NaN  NaN         NaN    fl  
3  NaN  NaN         NaN    ma  
4  NaN  NaN         NaN    nc

In [6]:
# Print missing value counts
df.isnull().sum()

id                   0
region               0
price                0
year              1205
manufacturer     17646
model             5277
condition       174104
cylinders       177678
fuel              3013
odometer          4400
title_status      8242
transmission      2556
VIN             161042
drive           130567
size            306361
type             92858
paint_color     130203
state                0
dtype: int64

In [7]:
# Remove all rows with values missing
df.dropna(inplace=True)

# Remove all the rows with price is 0
df.drop(df[df['price'] == 0].index, inplace=True)

In [8]:
# Print new dataframe
df.head()

id      region  price    year   manufacturer       model  \
215  7316482063  birmingham   4000  2002.0         toyota        echo   
219  7316429417  birmingham   2500  1995.0            bmw        525i   
268  7315946365  birmingham   9000  2008.0          mazda  miata mx-5   
337  7315354216  birmingham   8950  2011.0           ford       f-150   
338  7315349921  birmingham   4000  1972.0  mercedes-benz        benz   

     condition    cylinders fuel  odometer title_status transmission  \
215  excellent  4 cylinders  gas  155000.0        clean    automatic   
219       fair  6 cylinders  gas  110661.0        clean    automatic   
268  excellent  4 cylinders  gas   56700.0        clean    automatic   
337  excellent  6 cylinders  gas  164000.0        clean    automatic   
338       fair  6 cylinders  gas   88100.0        clean    automatic   

                   VIN drive       size         type paint_color state  
215  JTDBT123520243495   fwd    compact        sedan        blue    al  
219  WBAHD6322SGK86772   rwd   mid-size        sedan       white    al  
268  JM1NC25F570128735   rwd    compact  convertible       white    al  
337  1FTMF1CM6BKD55331   fwd  full-size        truck       white    al  
338        1.14023E+13   rwd  full-size        coupe      silver    al

In [9]:
# Print correlation matrix for numeric columns
df['year'] = df['year'].apply(lambda x: int(x)) # convert year to integer
df['cylinders'] = df['cylinders'].apply(lambda x: x.split()[0]) # convert cylinders to number
df['cylinders'] = df['cylinders'].apply(lambda x: x if x.isdigit() else 0) # fix non-number value
df[['price', 'year', 'odometer', 'cylinders']].corr()

price      year  odometer  cylinders
price      1.000000  0.315430 -0.233358   0.339627
year       0.315430  1.000000 -0.210172  -0.127261
odometer  -0.233358 -0.210172  1.000000   0.062923
cylinders  0.339627 -0.127261  0.062923   1.000000

In [10]:
# Unique values
print("condition = ", df['condition'].unique())
print("type = ", df['type'].unique())
print("transimission = ", df['transmission'].unique())
print("manufacturer = ", df['manufacturer'].unique())
print("drive = ", df['drive'].unique())
print("transmission = ", df['transmission'].unique())
print("fuel = ", df['fuel'].unique())



condition =  ['excellent' 'fair' 'good' 'like new' 'new' 'salvage']
type =  ['sedan' 'convertible' 'truck' 'coupe' 'SUV' 'hatchback' 'wagon' 'pickup'
 'other' 'van' 'mini-van' 'offroad' 'bus']
transimission =  ['automatic' 'manual' 'other']
manufacturer =  ['toyota' 'bmw' 'mazda' 'ford' 'mercedes-benz' 'ferrari' 'lexus'
 'chevrolet' 'audi' 'mitsubishi' 'infiniti' 'nissan' 'volkswagen' 'honda'
 'ram' 'hyundai' 'fiat' 'cadillac' 'jeep' 'gmc' 'jaguar' 'volvo' 'dodge'
 'kia' 'chrysler' 'subaru' 'acura' 'mini' 'buick' 'rover' 'lincoln'
 'porsche' 'saturn' 'harley-davidson' 'pontiac' 'mercury' 'tesla'
 'alfa-romeo' 'datsun' 'land rover' 'aston-martin']
drive =  ['fwd' 'rwd' '4wd']
transmission =  ['automatic' 'manual' 'other']
fuel =  ['gas' 'hybrid' 'diesel' 'other' 'electric']


### Data Understanding Summary

Price is not related to following features:
- VIN, id, region, model, fuel, drive, transmission, paint_color, state
  
Price is directly related to following numeric features:
- Year
- No. of Cylinders

Price is inversely related to followint numeric features:
- Odometer

### Data Preparation

After our initial exploration and fine-tuning of the business understanding, it is time to construct our final dataset prior to modeling.  Here, we want to make sure to handle any integrity issues and cleaning, the engineering of new features, any transformations that we believe should happen (scaling, logarithms, normalization, etc.), and general preparation for modeling with `sklearn`. 

In [11]:
# Drop columns not relevant to price like VIN, id, region, model, fuel, drive, paint_color, state
df = pd.read_csv("data/vehicles.csv")

# Drop all missing value rows to keep things simple
df = df.dropna()

# Print sample data
# df.info()

# Full feature list
feature_list = list(df.drop('price', axis=1).columns)
print("feature list = ", feature_list)

# Features to drop as no relationship to price
drop_list = ['VIN', 'id', 'region', 'model', 'paint_color', 'state']
df = df.drop(drop_list, axis=1)

# convert year to integer
df['year'] = df['year'].apply(lambda x: int(x)) 

# convert odometer to integer
df['odometer'] = df['odometer'].apply(lambda x: int(x)) 


# Convert cylinders column to integer
df['cylinders'] = df['cylinders'].apply(lambda x: x.split()[0]) # convert cylinders to number
df['cylinders'] = df['cylinders'].apply(lambda x: x if x.isdigit() else 0) # fix non-number value

# Convert features to one hot encoding
hot_encode_features = ['condition', 'manufacturer', 'title_status', 'type', 'size', 'drive', 'transmission', 'fuel']

for feature in hot_encode_features:
    df = pd.get_dummies(df, columns=[feature], prefix=feature)

# Verify numeric feature list
print("numeric features = ", set(feature_list) - set(drop_list + hot_encode_features))

# Remove 0 price rows
df = df.drop(df[df['price'] == 0].index)

feature list =  ['id', 'region', 'year', 'manufacturer', 'model', 'condition', 'cylinders', 'fuel', 'odometer', 'title_status', 'transmission', 'VIN', 'drive', 'size', 'type', 'paint_color', 'state']
numeric features =  {'year', 'odometer', 'cylinders'}


In [12]:
# Print sample rows
df.head()

price  year cylinders  odometer  condition_excellent  condition_fair  \
215   4000  2002         4    155000                 True           False   
219   2500  1995         6    110661                False            True   
268   9000  2008         4     56700                 True           False   
337   8950  2011         6    164000                 True           False   
338   4000  1972         6     88100                False            True   

     condition_good  condition_like new  condition_new  condition_salvage  \
215           False               False          False              False   
219           False               False          False              False   
268           False               False          False              False   
337           False               False          False              False   
338           False               False          False              False   

     ...  drive_fwd  drive_rwd  transmission_automatic  transmission_manual  \
215  ...       True      False                    True                False   
219  ...      False       True                    True                False   
268  ...      False       True                    True                False   
337  ...       True      False                    True                False   
338  ...      False       True                    True                False   

     transmission_other  fuel_diesel  fuel_electric  fuel_gas  fuel_hybrid  \
215               False        False          False      True        False   
219               False        False          False      True        False   
268               False        False          False      True        False   
337               False        False          False      True        False   
338               False        False          False      True        False   

     fuel_other  
215       False  
219       False  
268       False  
337       False  
338       False  

[5 rows x 85 columns]

In [13]:
# Print final info
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 32496 entries, 215 to 426833
Data columns (total 85 columns):
 #   Column                        Non-Null Count  Dtype 
---  ------                        --------------  ----- 
 0   price                         32496 non-null  int64 
 1   year                          32496 non-null  int64 
 2   cylinders                     32496 non-null  object
 3   odometer                      32496 non-null  int64 
 4   condition_excellent           32496 non-null  bool  
 5   condition_fair                32496 non-null  bool  
 6   condition_good                32496 non-null  bool  
 7   condition_like new            32496 non-null  bool  
 8   condition_new                 32496 non-null  bool  
 9   condition_salvage             32496 non-null  bool  
 10  manufacturer_acura            32496 non-null  bool  
 11  manufacturer_alfa-romeo       32496 non-null  bool  
 12  manufacturer_aston-martin     32496 non-null  bool  
 13  manufacturer_audi 

### Modeling

With your (almost?) final dataset in hand, it is now time to build some models.  Here, you should build a number of different regression models with the price as the target.  In building your models, you should explore different parameters and be sure to cross-validate your findings.

In [15]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.ensemble import RandomForestRegressor
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score

In [16]:
# Prepare training and test data
y = df[['price']]
X = df.drop(columns=['price'])

X_train, X_test, y_train, y_test = train_test_split(X, y)

print("X_train shape = ", X_train.shape)
print("y_train shape = ", y_train.shape)
print("X_test shape = ", X_test.shape)
print("y_test shape = ", y_test.shape)


X_train shape =  (24372, 84)
y_train shape =  (24372, 1)
X_test shape =  (8124, 84)
y_test shape =  (8124, 1)


In [17]:
# Create Linear regression model
model = LinearRegression()
model.fit(X_train, y_train)

# Prediction and errors
print("Linear regression model performance")
y_pred = model.predict(X_test)

mae = mean_absolute_error(y_pred, y_test)
print("mae = ", mae)

mse = mean_squared_error(y_pred, y_test)
print("mse = ", mse)

score = r2_score(y_pred, y_test)
print("r2 score = ", score)

Linear regression model performance
mae =  5862.932149109358
mse =  79151496.1861757
r2 score =  0.10741872277893671


In [18]:
# Create Random forest regressor
model = RandomForestRegressor(n_estimators=100, random_state=42)
model.fit(X_train, y_train)

# Prediction and errors
print("Random forest model performance")
y_pred = model.predict(X_test)

mae = mean_absolute_error(y_pred, y_test)
print("mae = ", mae)

mse = mean_squared_error(y_pred, y_test)
print("mse = ", mse)

score = r2_score(y_test, y_pred)
print("r2 score = ", score)

/opt/miniconda3/envs/berkeley_ai/lib/python3.11/site-packages/sklearn/base.py:1473: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  return fit_method(estimator, *args, **kwargs)


Random forest model performance
mae =  2157.8001739574124
mse =  22905083.51821056
r2 score =  0.8704623425645595


### Evaluation

With some modeling accomplished, we aim to reflect on what we identify as a high-quality model and what we are able to learn from this.  We should review our business objective and explore how well we can provide meaningful insight into drivers of used car prices.  Your goal now is to distill your findings and determine whether the earlier phases need revisitation and adjustment or if you have information of value to bring back to your client.

Here is the summary of different model evaluated

| Model Type | R2 Score |
| -----------|----------|
| Linear Regression | 0.10 |
| Random Forest | 0.87 |

#### Based on above observation, Random forest based model performs better

### Deployment

Now that we've settled on our models and findings, it is time to deliver the information to the client.  You should organize your work as a basic report that details your primary findings.  Keep in mind that your audience is a group of used car dealers interested in fine-tuning their inventory.